### ISSUE TITLE
new loaders: move them out of v0

### ISSUE THREAD

**Issue by aghassemi** [ampproject/amphtml#23108]
We should be able to move majority of the new loader code to a new extensions and gets dynamically loaded. We may only need to keep the code that creates the container and decided whether the gray placeholder is needed in v0.

This is because new loaders don't show up for 0.6 seconds anyway, so if we dynamically load the extension and then adjust the start time of animation as `0.6 -  extension load time`, it will have no effect in vas majority of the cases. Worse case the loaders take a bit more than 0.6s to show up which is worth the savings.

**Comment by sparhami** [comment]
# Background

#20237 for the loaders feature.

The code for displaying the loaders has traditionally lived in AMP. This has not been a major problem since the code has been relatively small and asynchronously loading it would likely provide little benefit when considering the extra logic to asynchronously load the code itself.

Unlike the existing loader, the new loader delays showing up by 600ms (this time may be adjusted in the future based on data that we gather). This reduces the importance of having the loaders code available immediately.

# Potential Benefit

By moving the new loaders code out of the runtime, we can achieve a 1.45KB reduction in payload on the main binary. This represents a 1.7% reduction in the binary size. Once the new loaders have launched and we remove the old loaders code, we can reduce the payload by a total of  1.83KB, or 2.1% of the current binary size.

# Design

Create an asynchronously, always loaded piece of JS to render loaders. This would be created as an extension, similiar to how `amp-auto-lightbox` works. Ideally, we would have a different mechanism for async loading than an extension. This code does not behave the same as other extensions, and there is no use for page developers.

The basics are of the approach are:

1. Create an extension
    1. Register a `Service` for creating the loader DOM
    1. Insert the necessary CSS for the current doc
1. Load the extension and retrieve the service

The code for this would look like:


```javacript
AMP.extension('amp-loader', '0.1', AMP => {
  AMP.registerServiceForDoc('loader', LoaderService);
  Services.extensionsFor(AMP.win).addDocFactory(ampDoc => {
    installStylesForDoc(ampDoc, CSS, () => {}, false, 'amp-loader');
  });
});
```
Note that since we do not have a tagname / custom element associated with the loader, we directly install the styles. In the future, we may want to make the loader itself a proper element.

```javascript
Services.extensionsFor(ampDoc.win)
  .installExtensionForDoc(ampDoc, 'amp-loader')
  .then(() => Services.loaderServiceForDoc(element));
```

## Handling loading delay

Since the code is loaded asynchronously, we can subtract the loading delay from the delay for showing the loader. This will result in the loader showing up at the same time it normally would in some (many? most?) cases.


# Alternatives Considered

## Preload the amp-loader code

One option would be to initiate a preload the code for amp-loader, either early on in the binary, or through an optimization through the AMP cache. This would make the loader code be more likely to be available when needed, especially for slower network connections.

This approach is not taken for now, as loading the code earlier would contend for badwidth with extensions, images and other resources. In addition, the JS execution could block other more important things from running.

We may want to try this approach in the future and collect metrics to see what the impact of preloading would be fore the rest of the page layout.

## Create the loader as a custom element

The loader itself could be created as a custom element, removing the need for registering/retrieving a service. Since elements have a cost associated with them and their lifecycle, the current implementation instead directly creates the loader DOM structure. The loader could potentially be created as a plain custom element (and not an AMP element), but this would require additional research.
